// @cerbos/embedded/lib/protobuf/cerbos/engine/v1/engine.js
"use strict";
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Principal_AttrEntry = exports.Principal = exports.Resource_AttrEntry = exports.Resource = exports.OutputEntry = exports.PlanResourcesFilter_Expression_Operand = exports.PlanResourcesFilter_Expression = exports.PlanResourcesFilter = exports.PlanResourcesInput_Resource_AttrEntry = exports.PlanResourcesInput_Resource = exports.PlanResourcesInput = exports.planResourcesFilter_KindToJSON = exports.planResourcesFilter_KindFromJSON = exports.PlanResourcesFilter_Kind = exports.protobufPackage = void 0;
exports.protobufPackage = "cerbos.engine.v1";
var PlanResourcesFilter_Kind;
(function (PlanResourcesFilter_Kind) {
    PlanResourcesFilter_Kind[PlanResourcesFilter_Kind["KIND_UNSPECIFIED"] = 0] = "KIND_UNSPECIFIED";
    PlanResourcesFilter_Kind[PlanResourcesFilter_Kind["KIND_ALWAYS_ALLOWED"] = 1] = "KIND_ALWAYS_ALLOWED";
    PlanResourcesFilter_Kind[PlanResourcesFilter_Kind["KIND_ALWAYS_DENIED"] = 2] = "KIND_ALWAYS_DENIED";
    PlanResourcesFilter_Kind[PlanResourcesFilter_Kind["KIND_CONDITIONAL"] = 3] = "KIND_CONDITIONAL";
})(PlanResourcesFilter_Kind || (exports.PlanResourcesFilter_Kind = PlanResourcesFilter_Kind = {}));
function planResourcesFilter_KindFromJSON(object) {
    switch (object) {
        case 0:
        case "KIND_UNSPECIFIED":
            return PlanResourcesFilter_Kind.KIND_UNSPECIFIED;
        case 1:
        case "KIND_ALWAYS_ALLOWED":
            return PlanResourcesFilter_Kind.KIND_ALWAYS_ALLOWED;
        case 2:
        case "KIND_ALWAYS_DENIED":
            return PlanResourcesFilter_Kind.KIND_ALWAYS_DENIED;
        case 3:
        case "KIND_CONDITIONAL":
            return PlanResourcesFilter_Kind.KIND_CONDITIONAL;
        default:
            throw new globalThis.Error("Unrecognized enum value " +
                object +
                " for enum PlanResourcesFilter_Kind");
    }
}
exports.planResourcesFilter_KindFromJSON = planResourcesFilter_KindFromJSON;
function planResourcesFilter_KindToJSON(object) {
    switch (object) {
        case PlanResourcesFilter_Kind.KIND_UNSPECIFIED:
            return "KIND_UNSPECIFIED";
        case PlanResourcesFilter_Kind.KIND_ALWAYS_ALLOWED:
            return "KIND_ALWAYS_ALLOWED";
        case PlanResourcesFilter_Kind.KIND_ALWAYS_DENIED:
            return "KIND_ALWAYS_DENIED";
        case PlanResourcesFilter_Kind.KIND_CONDITIONAL:
            return "KIND_CONDITIONAL";
        default:
            throw new globalThis.Error("Unrecognized enum value " +
                object +
                " for enum PlanResourcesFilter_Kind");
    }
}
exports.planResourcesFilter_KindToJSON = planResourcesFilter_KindToJSON;
exports.PlanResourcesInput = {
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
};
exports.PlanResourcesInput_Resource = {
    fromJSON(object) {
        return {
            kind: isSet(object.kind) ? globalThis.String(object.kind) : "",
            attr: isObject(object.attr)
                ? Object.entries(object.attr).reduce((acc, [key, value]) => {
                    acc[key] = value;
                    return acc;
                }, {})
                : {},
            policyVersion: isSet(object.policyVersion)
                ? globalThis.String(object.policyVersion)
                : "",
            scope: isSet(object.scope) ? globalThis.String(object.scope) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.kind !== "") {
            obj.kind = message.kind;
        }
        if (message.attr) {
            const entries = Object.entries(message.attr);
            if (entries.length > 0) {
                obj.attr = {};
                entries.forEach(([k, v]) => {
                    obj.attr[k] = v;
                });
            }
        }
        if (message.policyVersion !== "") {
            obj.policyVersion = message.policyVersion;
        }
        if (message.scope !== "") {
            obj.scope = message.scope;
        }
        return obj;
    },
};
exports.PlanResourcesInput_Resource_AttrEntry = {
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object?.value) ? object.value : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== undefined) {
            obj.value = message.value;
        }
        return obj;
    },
};
exports.PlanResourcesFilter = {
    fromJSON(object) {
        return {
            kind: isSet(object.kind)
                ? planResourcesFilter_KindFromJSON(object.kind)
                : 0,
            condition: isSet(object.condition)
                ? exports.PlanResourcesFilter_Expression_Operand.fromJSON(object.condition)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.kind !== 0) {
            obj.kind = planResourcesFilter_KindToJSON(message.kind);
        }
        if (message.condition !== undefined) {
            obj.condition = exports.PlanResourcesFilter_Expression_Operand.toJSON(message.condition);
        }
        return obj;
    },
};
exports.PlanResourcesFilter_Expression = {
    fromJSON(object) {
        return {
            operator: isSet(object.operator)
                ? globalThis.String(object.operator)
                : "",
            operands: globalThis.Array.isArray(object?.operands)
                ? object.operands.map((e) => exports.PlanResourcesFilter_Expression_Operand.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.operator !== "") {
            obj.operator = message.operator;
        }
        if (message.operands?.length) {
            obj.operands = message.operands.map((e) => exports.PlanResourcesFilter_Expression_Operand.toJSON(e));
        }
        return obj;
    },
};
exports.PlanResourcesFilter_Expression_Operand = {
    fromJSON(object) {
        return {
            node: isSet(object.value)
                ? { $case: "value", value: object.value }
                : isSet(object.expression)
                    ? {
                        $case: "expression",
                        expression: exports.PlanResourcesFilter_Expression.fromJSON(object.expression),
                    }
                    : isSet(object.variable)
                        ? {
                            $case: "variable",
                            variable: globalThis.String(object.variable),
                        }
                        : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.node?.$case === "value") {
            obj.value = message.node.value;
        }
        if (message.node?.$case === "expression") {
            obj.expression = exports.PlanResourcesFilter_Expression.toJSON(message.node.expression);
        }
        if (message.node?.$case === "variable") {
            obj.variable = message.node.variable;
        }
        return obj;
    },
};
exports.OutputEntry = {
    fromJSON(object) {
        return {
            src: isSet(object.src) ? globalThis.String(object.src) : "",
            val: isSet(object?.val) ? object.val : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.src !== "") {
            obj.src = message.src;
        }
        if (message.val !== undefined) {
            obj.val = message.val;
        }
        return obj;
    },
};
exports.Resource = {
    fromJSON(object) {
        return {
            kind: isSet(object.kind) ? globalThis.String(object.kind) : "",
            policyVersion: isSet(object.policyVersion)
                ? globalThis.String(object.policyVersion)
                : "",
            id: isSet(object.id) ? globalThis.String(object.id) : "",
            attr: isObject(object.attr)
                ? Object.entries(object.attr).reduce((acc, [key, value]) => {
                    acc[key] = value;
                    return acc;
                }, {})
                : {},
            scope: isSet(object.scope) ? globalThis.String(object.scope) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.kind !== "") {
            obj.kind = message.kind;
        }
        if (message.policyVersion !== "") {
            obj.policyVersion = message.policyVersion;
        }
        if (message.id !== "") {
            obj.id = message.id;
        }
        if (message.attr) {
            const entries = Object.entries(message.attr);
            if (entries.length > 0) {
                obj.attr = {};
                entries.forEach(([k, v]) => {
                    obj.attr[k] = v;
                });
            }
        }
        if (message.scope !== "") {
            obj.scope = message.scope;
        }
        return obj;
    },
};
exports.Resource_AttrEntry = {
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object?.value) ? object.value : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== undefined) {
            obj.value = message.value;
        }
        return obj;
    },
};
exports.Principal = {
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.String(object.id) : "",
            policyVersion: isSet(object.policyVersion)
                ? globalThis.String(object.policyVersion)
                : "",
            roles: globalThis.Array.isArray(object?.roles)
                ? object.roles.map((e) => globalThis.String(e))
                : [],
            attr: isObject(object.attr)
                ? Object.entries(object.attr).reduce((acc, [key, value]) => {
                    acc[key] = value;
                    return acc;
                }, {})
                : {},
            scope: isSet(object.scope) ? globalThis.String(object.scope) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== "") {
            obj.id = message.id;
        }
        if (message.policyVersion !== "") {
            obj.policyVersion = message.policyVersion;
        }
        if (message.roles?.length) {
            obj.roles = message.roles;
        }
        if (message.attr) {
            const entries = Object.entries(message.attr);
            if (entries.length > 0) {
                obj.attr = {};
                entries.forEach(([k, v]) => {
                    obj.attr[k] = v;
                });
            }
        }
        if (message.scope !== "") {
            obj.scope = message.scope;
        }
        return obj;
    },
};
exports.Principal_AttrEntry = {
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object?.value) ? object.value : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== undefined) {
            obj.value = message.value;
        }
        return obj;
    },
};
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
